src/universe/visualization/img_tools.py

```diff
@@ -14,3 +18,313 @@ def rotate(img: np.ndarray, angle: float) -> np.ndarray:
+        result = cv.cvtColor(self._original.copy(), cv.COLOR_BGR2RGB)
+        # convert the image to grayscale if necessary
+        if self.grayscale:
+            result = grayscale_image(result)
+        # darken the image if necessary
+        if self._darkness > 0:
+            result = darken_image(result, self._darkness)
+        # draw the layers on the image
+        for layer in self.layers:
+            result = layer.draw_on(result)
+        # resize the image if necessary
+        if self.width != self._original.shape[1] or self.height != self._original.shape[0]:
+            result = cv.resize(result, (self.width, self.height))
+        return result
+
+    def plot(self, title: str = None, axis: bool = False):
+        """
+        Shows a matplotlib plot of the image with the given title. Also shows the legend.
+
+        :param title: title of the plot
+        :param axis: whether to show the axis or not
+        """
+
+        # add legend to plot
+        legend = []
+        for layer in self.layers:
+            if layer.name:
+                legend.append(mpatches.Patch(color=[x / 255 for x in layer.color], label=layer.name))
+        if len(legend) > 0:
+            plt.legend(handles=legend, loc='lower right')
+
+        if title:
+            plt.title(title)
+        if not axis:
+            plt.axis('off')
+        # show plot
+        plt.imshow(self.image)
+
+    def save(self, path: str) -> bool:
+        """
+        Saves the image with the layers drawn on it to the given path.
+
+        :param path: path to save the image to
+        :return: whether the image was saved successfully or not
+        """
+        return cv.imwrite(path, cv.cvtColor(self.image, cv.COLOR_RGB2BGR))
+
+    def resize(self, width: int = None, height: int = None):
+        """
+        Resizes the image to the given width and height.
+
+        :param width: new width of the image
+        :param height: new height of the image
+        """
+        if width:
+            assert width > 0, 'Width must be greater than 0'
+            self.width = width
+        if height:
+            assert height > 0, 'Height must be greater than 0'
+            self.height = height
+
+    def scale(self, multiplier: float):
+        """
+        Scales the image by the given multiplier.
+
+        :param multiplier: multiplier to scale the image by
+        """
+        assert multiplier > 0, 'Multiplier must be greater than 0'
+        self.width = round(self.width * multiplier)
+        self.height = round(self.height * multiplier)
+
+
+def grayscale_image(image: cv.Mat) -> cv.Mat:
+    """
+    Converts the given image to grayscale and then to the original color space.
+
+    :param image: image to convert
+    :return: grayscale image
+    """
+    assert image is not None, 'Image does not exist'
+    return cv.cvtColor(cv.cvtColor(image, cv.COLOR_RGB2GRAY), cv.COLOR_GRAY2RGB)
+
+
+def darken_image(image: cv.Mat, darkness: int = 50) -> cv.Mat:
```

Comment:
wouldn't it be more convenient to enforce the darkness between 0 and 1 (int -> float)? That way you wouldn't need to divide it by 100.

Reply:
```suggestion
def darken_image(image: cv.Mat, darkness: float = .5) -> cv.Mat:
```

Reply:
Ok, but then can we still call it a percentage, or should we call it something else?

Reply:
the range 0..1 is definitely can be a percentage